streaming/in.go

```diff
@@ -47,315 +87,442 @@ func newCrypt4GHInternalReader(reader io.Reader, readerPrivateKey [chacha20poly1
 		if err != nil {
-			return 0, err
+			return nil, err
 		}
 	}
 
-	return c.buffer.Read(p)
+	return &crypt4GHInternalReader, nil
 }
 
-func (c *crypt4GHInternalReader) ReadByte() (byte, error) {
+// ensureBuffer ensure the decrypted buffer is not empty
+// unless we have reached the end of the stream
+// it also makes sure any remaining data in the buffer
+// matches what should be observed in the consumer stream
+// at c.streamPos, specifically a Read or ReadByte will
+// return the data that should be seen at c.streamPos and
+// forward
+//
+// Returns any error encountered.
+func (c *crypt4GHInternalReader) ensureBuffer() (err error) {
+
+	neededSegment, err := c.consumerOffsetToSegment(c.streamPos)
+	if err != nil {
+		// Outside of file? Forward error (EOF)
+		return err
+	}
+
+	neededPos, err := c.consumerOffsetToEncryptedStreamOffset(c.streamPos)
+	if err != nil {
+		return err
+	}
+
+	// Figure out the needed offset within the segment
+	segmentOffset := int(neededPos - neededSegment*int64(headers.UnencryptedDataSegmentSize))
+	bufferOffset := c.bufferUse - c.buffer.Len()
+
+	if c.lastDecryptedSegment != neededSegment || segmentOffset < bufferOffset {
+		// If we want to read another segment than the current or if we've
+		// already read past the desired offset, we need to fetch data, signal
+		// this by throwing away whatever we currently have
+		c.buffer.Reset()
+	}
+
+	// If we don't have any data on hand, fetch more
 	if c.buffer.Len() == 0 {
-		err := c.fillBuffer()
-		if err != nil {
-			return 0, err
+		if err := c.fillBuffer(); err != nil {
+			return err
 		}
+		bufferOffset = 0
 	}
 
-	return c.buffer.ReadByte()
+	// Find the correct place in the buffert
+	if bufferOffset < segmentOffset {
+		toSkip := int(segmentOffset) - bufferOffset
+		_ = c.buffer.Next(toSkip)
+	}
+
+	return nil
+}
+
+// consumerOffsetToSegment returns the segment in the underlying stream
+// for the passed consumer offset.
+//
+// It can possibly return error EOF if the passed offset is outside
+// of the exposed stream, but this is not guaranteed
+func (c *crypt4GHInternalReader) consumerOffsetToSegment(n int64) (int64, error) {
+	// Figure out the segment
+
+	eo, err := c.consumerOffsetToEncryptedStreamOffset(n)
+	segment := eo / int64(headers.UnencryptedDataSegmentSize)
+
+	return segment, err
 }
 
-func (c *crypt4GHInternalReader) Discard(n int) (discarded int, err error) {
-	if n < 0 {
-		return
+// consumerOffsetToSEncryptedStreamOffset returns the offset in the underlying
+// stream for the passed consumer offset. This is excluding the extra bytes
+// added by the crypt4gh file format (e.g. headers or segment nonce and mac).
+//
+// It can possibly return error EOF if the passed offset is outside of the
+// exposed stream, but this is not guaranteed.
+func (c *crypt4GHInternalReader) consumerOffsetToEncryptedStreamOffset(n int64) (int64, error) {
+	// Calculate the offset in the encrypted stream from the consumer visible
+	// offset
+	//
```

Comment:
extra comment line?

not the only one, not sure why though, can we clean those up ?

Reply:
These are intentional to give air and more specifically to separate paragraphs (in godoc in theory but since these are not public, mostly in VS Code hoovering "tips").

OK?